merlin/spec/specification.py

```diff
@@ -354,8 +374,23 @@ def process_spec_defaults(self):
         if self.merlin["resources"]["workers"] is None:
             self.merlin["resources"]["workers"] = {"default_worker": defaults.WORKER}
         else:
-            for worker, vals in self.merlin["resources"]["workers"].items():
+            # Gather a list of step names defined in the study
+            steps = self.get_study_step_names()
+            modified_steps = False
+            # Loop through each worker and fill in the defaults
+            for _, vals in self.merlin["resources"]["workers"].items():
                 MerlinSpec.fill_missing_defaults(vals, defaults.WORKER)
+                # Loop through the list of steps assigned to a worker
+                for step in vals["steps"]:
+                    # If the assigned step is still in the list of existing steps, remove it
+                    if step in steps:
+                        modified_steps = True
+                        steps.remove(step)
+            # If other steps have been assigned a worker but there are still steps remaining
+            # that haven't been assigned a worker yet, assign the remaining steps to the default worker
+            if modified_steps and steps:
```

Comment:
You could also get fancy and condense a lot of this looping into the the subtraction of various sets. 

steps_that_need_workers = set(workfow_steps) - set(worker_steps)

If you want to be fancy 🎩

Reply:
I just did this, ran the test suite, and now I remember why I had the boolean there. 

The current logic is that if there are no workers defined in a spec then by default the `default_worker` is created and it has `steps: ["all"]` in it's worker settings. With the new code I'm adding, instead of it being `steps: ["all"]` it would be `steps: [<list of every step in workflow>]` which causes a bunch of tests to fail. 

Rather than having a bool to keep track of, I think I could just change the if statement to be something like `if steps_that_need_workers and (steps_that_need_workers != all_workflow_steps)`, where `all_workflow_steps` is a list of all the steps in the workflow (essentially `self.get_study_step_names()`).

Reply:
I just now saw your comment on using sets, I think that would be cleaner than a lot of this so maybe ignore the above message for now while I check that out.

Reply:
Using sets is a lot cleaner but I still needed that if check that I was talking about. I just pushed the new changes if you want to see them.

Reply:
Nice